File: Trees/7. Lowest Common Ancestor of a Binary Search Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	var pv = p.Val
	var qv = q.Val
	var rv = root.Val
	if pv > rv && qv > rv {
		return lowestCommonAncestor(root.Right, p, q)
	} else if pv < rv && qv < rv {
		return lowestCommonAncestor(root.Left, p, q)
	} else if pv == rv {
		return p
	} else if qv == rv {
		return q
	} else {
		return root
	}
}

File: Trees/7. Lowest Common Ancestor of a Binary Search Tree/solution.scala
/** Definition for a binary tree node. class TreeNode(var _value: Int) { var
  * value: Int = _value var left: TreeNode = null var right: TreeNode = null }
  */

object Solution {
  def lowestCommonAncestor(
      root: TreeNode,
      p: TreeNode,
      q: TreeNode
  ): TreeNode = {
    val rv = root.value
    (p.value, q.value) match {
      case (pv, qv) if pv > rv && qv > rv =>
        lowestCommonAncestor(root.right, p, q)
      case (pv, qv) if pv < rv && qv < rv =>
        lowestCommonAncestor(root.left, p, q)
      case (pv, _) if pv == rv => p
      case (_, qv) if qv == rv => q
      case (_, _)              => root
    }
  }
}
